internal/nginx/config/servers.go

```diff
@@ -308,6 +318,76 @@ func createMatchLocation(path string) http.Location {
 	}
 }
 
+func generateProxySetHeaders(filters *v1beta1.HTTPHeaderFilter) ([]http.Header, []http.Map) {
+	var proxySetHeaders []http.Header
+	var proxySetHeadersMaps []http.Map
+	if filters != nil {
+		if len(filters.Add) > 0 {
+			addHeaders, headersMap := convertAddHeaders(filters.Add)
+			proxySetHeaders = append(proxySetHeaders, addHeaders...)
+			proxySetHeadersMaps = append(proxySetHeadersMaps, headersMap...)
+		}
+		if len(filters.Set) > 0 {
+			setHeaders := convertSetHeaders(filters.Set)
+			proxySetHeaders = append(proxySetHeaders, setHeaders...)
+		}
+		// If the value of a header field is an empty string then this field will not be passed to a proxied server
+		for _, h := range filters.Remove {
+			proxySetHeaders = append(proxySetHeaders, http.Header{
+				Name:  h,
+				Value: "",
+			})
+		}
+	}
+	return proxySetHeaders, proxySetHeadersMaps
+}
+
+func convertAddHeaders(headers []v1beta1.HTTPHeader) ([]http.Header, []http.Map) {
+	var locHeaders []http.Header
+	var headerMaps []http.Map
+	for _, h := range headers {
+		headerMap, varName := createHeadersMap(string(h.Name))
+		headerMaps = append(headerMaps, headerMap)
+		locHeaders = append(locHeaders, http.Header{
+			Name:  string(h.Name),
+			Value: "${" + varName + "}" + h.Value,
+		})
+	}
+	return locHeaders, headerMaps
+}
+
+func createHeadersMap(name string) (http.Map, string) {
+	underscoreName := strings.ToLower(strings.ReplaceAll(name, "-", "_"))
+	httpVarSource := "${http_" + underscoreName + "}"
+	mapVarName := underscoreName + "_header_var"
```

Comment:
we need to make sure the generated name:
- unique among all HTTPRoutes - to prevent NGINX reload failures
- is not too long - too prevent cases like https://github.com/nginxinc/kubernetes-ingress/issues/2802